chemprop/data/utils.py

```diff
@@ -16,16 +16,31 @@
 from chemprop.features import load_features, load_valid_atom_features
 
 
-def preprocess_smiles_columns(smiles_columns: Optional[Union[str, List[Optional[str]]]]) -> List[Optional[str]]:
+def preprocess_smiles_columns(path: str,
+                              smiles_columns: Optional[Union[str, List[Optional[str]]]],
+                              number_of_molecules: int = 1) -> List[Optional[str]]:
```

Comment:
I have a high level idea for how this function could behave to avoid needing the list check in the util functions and the dummy file in the website. Let me know whether you think it would work.

(pseudocode)
```
if smiles_columns is None
    if path is a valid file
        return first number_of_molecules columns of file header
    else (e.g. `'None'` from the web view)
        return list of None of length number_of_molecules
else
    if smiles_columns is not a list
        wrap in list
    if path is a valid file
        check that smiles_columns exists in file header
    return smiles_columns
```

Reply:
This is a good structure. The valid file check seems a better way to address the dummy web input. And this is a good way to run if we want to always pass through it on the way through the utils.

My preference would be to make smiles_columns a required input to those functions with no default value and remove the `preprocess_smiles_columns` backstop from the utility functions entirely. But, always running through them maintains the flexibility well without changing the outside view of the utilities.

Reply:
Wait, actually there's still a problem. Your structure doesn't include the check for whether the length of the smiles_columns is the same as the number_of_molecules. We need that check in the initial arguments processing, when the input will be a list but no guarantees it's the right length.

But the different utils functions don't have access the args.number_of_molecules, so `preprocess_smiles_columns` will error out for a mismatch if we ever try to run with 2 molecules.

I guess there are three options: 1) Stay with the existing structure with awkward type checks in the utils to detect if it needs to pass through the function again. 2) Make number_of_molecules an input to all the associated utils and always pass through the function. 3) Make a list input for smiles_columns required for the utils and remove the function check so it never passes through the function.

Reply:
Thought about this a little more. Making more information required (either smiles_columns or number_of_molecules) really does hamper the usefulness of general utilities. More information is required to either run through all the time or run through never. So I'm going to go with a similar check to what is there now, but paring it down to checking for `None` instead of being a list.

Reply:
Thanks for catching that omission. I think keeping the check in the utils makes sense then.

I do think it would be good to add the file check to avoid needing a dummy file when only parsing certain args manually.

Reply:
I am otherwise adopting your file structure. The file check is *much* better than making a dummy file just to read.